### ISSUE TITLE
What are the criteria for embedded content vs. links

### ISSUE THREAD

**Issue by petersilva** [wmo-im/GTStoWIS2#36]

**Comment by petersilva** [comment]
thought experiment picking ugly numbers for illustrative purposes:

If we have a maximum message size, then experience with WMO indicates the average message size will converge to about half of the maximum... (when it was 14K, we observed a message size of 7K on our links... but there was segmentation in the story... who knows how it will change in the new methods.)
If we pick 8M as a new max, then imagine 4M becomes the average. A message without payload is about
If we are transferring 100 messages per second.  without payload, assume a message is 512 bytes.
so at any given message transfer rate, the size difference is 8190:1 ... so if we can support
100 messages/second coming in over the message protocol, then we should get 0.02 messages/second
Or you need the message protocol to run 8000 times faster, or some combination.

It becomes a question of how good are brokers at doing actual data transfer (as opposed to switching.)
The traditional answer is: not very good.  But I fully get that people intuitively think it should be faster, and in many cases it should be, but in a lot of other common cases, I expect it to turn out to be counter-productive.

**Comment by remygiraud** [comment]
With my understanding, brokers are intended to work best with "small" messages. Everyone will agree that a 1GB file is not small. On the other end of the spectrum, 1kB is small. So, a range of 1 to 1000000 between the two. For brokers with the anticipated workload, can we consider 100kB, 1MB being "small"? I agree we need to run some experiment.

**Comment by petersilva** [comment]
So far, in the committee, we had proposed 4KB as a definition of "small enough to be embedded" and the strategy was just to embed everything smaller than that.  I wanted 512 bytes... but in the interest of consensus, we have been using 4KB.

I expect that CAP messages are typically in the 100KB to 500KB range, so about 50x-100x larger... I think the content agnostic approach (embedding all messages smaller than an embedding threshold) is no longer reasonable at that size, as it will slow things down unacceptably for a large number of cases.

We could get more sophisticated, and embed bigger messages as long as they are sufficiently rare and important, but that means we need to understand what we are sending, as opposed to being content agnostic as we have so far succeeded in being.

**Comment by eliot-christian** [comment]
In my experience, CAP messages are typically quite small.. My guess is that they average about 1,000 bytes (1 K). 

To my mind, sudden onset emergencies provide the primary Use Case for embedding CAP alerts in a message queue. In this case, seconds can be the difference between a life-saving alert and an alert that arrives too late. Examples of sudden-onset emergencies include earthquake-early-warning (alerts trying to outrace the earthquake wave propagation) and its analogues in tsunami, flash floods, landslides, volcanic eruption, space weather, et al. Also qualifying as sudden-onset emergencies are public safety matters such as 'active shooter' situations.

We should bear in mind that CAP alerts are often handled without human mediation, as in the triggering of sirens, traffic signals, bridge and tunnel gates, etc,

**Comment by petersilva** [comment]
You can see Canadian CAP here for the last few weeks: https://dd.weather.gc.ca/alerts/cap/

Looking at today, for one of our seven storm prediction centres, we have issued about a dozen warnings, most of them are in the 200KB to 400KB range.  about 1/3 are in the 18KB range.  None are smaller than that.

**Comment by petersilva** [comment]
On the other hand, if people are ok with messages < 4KB being embedded, that is the committee working hypothesis anyways, and if folks agree that that convention includes CAP (ie. we do not expect CAP message bigger than 4KB to be embedded.) then there is already complete agreement.

**Comment by eliot-christian** [comment]
You are correct that these are huge CAP alerts.

**Comment by amilan17** [comment]
decision in: wmo-im/wis2-notification-message#6

**Comment by eliot-christian** [comment]
I agree with those on this thread who assert that the link is a MUST in all
cases and inline content is a MAY.

From a law and policy perspective, the CAP alert as originally published is
the legal, public document that must be maintained as the permanent record
of what the alerting authority sent to the public for the emergency.
Therefore, the link to that CAP alert is of the essence and not merely a
convenience.

I draw attention in this regard to the new CEN Workshop Agreement:
"Requirements and recommendations for social media early warning messages
in crisis and disaster management". It includes this requirement: "To
ensure consistency, clarity and completeness of the messages disseminated
simultaneously across different channels, the social media early warning
messages and related notifications shall refer to the persistent unique URL
of the CAP message."

Eliot Christian
